crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/fetch/fetch_task.rs

```diff
@@ -0,0 +1,284 @@
+
+        // The item should get fetched
+        assert_eq!(1, fetched.iter().flatten().count());
+
+        // Simulate the requested item being sent back
+        held_op_data.lock().insert(test_key_hash(1));
+
+        // Move forwards by 5 minutes so that the item will be retried
+        tokio::time::advance(Duration::from_secs(5 * 60)).await;
+
+        // Then the item should be removed from the pool
+        wait_for_pool_n(&fetch_pool, 0).await;
+
+        internal_sender_test
+            .ghost_actor_shutdown_immediate()
+            .await
+            .unwrap();
+    }
+
+    #[ignore = "open question"]
+    #[tokio::test(start_paused = true)]
+    async fn fetch_single_region() {
+        let (_task, fetch_pool, internal_sender_test, _held_op_data) =
+            setup(InternalStub::new()).await;
+
+        fetch_pool.push(test_req_region(1, None, test_source(1)));
+        wait_for_pool_n(&fetch_pool, 1).await;
+
+        let fetched = wait_for_fetch_n(internal_sender_test.clone(), 1).await;
+
+        assert_eq!(1, fetched.iter().flatten().count());
+
+        // TODO No way to mark this as fetched?
+
+        // Move forwards by 5 minutes so that the item will be retried
+        tokio::time::advance(Duration::from_secs(5 * 60)).await;
+
+        // Never removed after fetched
+        wait_for_pool_n(&fetch_pool, 0).await;
+
+        internal_sender_test
+            .ghost_actor_shutdown_immediate()
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn fetch_task_shuts_down_if_internal_sender_closes() {
+        let (task, fetch_pool, internal_sender_test, _held_op_data) =
+            setup(InternalStub::new()).await;
+
+        // Do enough testing to prove the loop is up and running
+        fetch_pool.push(test_req_region(1, None, test_source(1)));
+        wait_for_pool_n(&fetch_pool, 1).await;
+        wait_for_fetch_n(internal_sender_test.clone(), 1).await;
+
+        // Shut down ghost actor to close the sender
+        internal_sender_test
+            .ghost_actor_shutdown_immediate()
+            .await
+            .unwrap();
+
+        // Move forwards by 5 minutes so that the item will be retried
+        tokio::time::advance(Duration::from_secs(5 * 60)).await;
+
+        tokio::time::timeout(Duration::from_secs(5), {
+            let task = task.clone();
+            async move {
+                loop {
+                    if task.read().is_finished {
+                        return;
+                    }
+
+                    tokio::time::sleep(Duration::from_millis(1)).await;
+                }
+            }
+        })
+        .await
+        .expect("Task should have shut down but is reporting that it is still running");
+
+        assert!(task.read().is_finished);
+    }
+
+    // TODO the API supports batch queries, why not query in batch? We are pushing extra requests through a bottleneck
```

Comment:
This is a relatively significant performance improvement that we could make quite easily

Reply:
For clarity, you are talking about batching the `check_op_data` call to the local host so we don't have to capture separate DB locks?

I agree that is worth doing.

Separately, I think your `1 here and still 3 below` comment below is incorrect and they should all stay `3`s, since `wait_for_fetch_n` is waiting on the internal sender, which should still include three separate fetch calls to remotes?

Let me know if I'm misunderstanding something.

Reply:
Yes to the first question. It's not just db locks though it's also extra messages going through the `p2p_event_task` task that is a bottleneck.

Yes you're right this comment isn't accurate. It's fewer calls to check_op_data. I'll correct the test and the comment here